### PR TITLE
netlink: introduce typesafe netlink functions

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

**Comment by evverx** [comment]
This pull request **introduces 1 alert** when merging d6f7843390cba42b65d1f3c22996c4d31d5fa3c0 into 09dad04c49cae3ad2b319c9b4e7773fedd34309a - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-c5fb581792ce605189def4f9279998f81ef4bf6e)

**new alerts:**

* 1 for Missing header guard

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/network/networkd-dhcp4.c [hunk 1]
```diff
@@ -216,8 +216,7 @@ static int dhcp_lease_lost(Link *link) {
                                         assert_se(sd_dhcp_route_get_destination(routes[i], &route->dst.in) >= 0);
                                         assert_se(sd_dhcp_route_get_destination_prefix_length(routes[i], &route->dst_prefixlen) >= 0);
 
-                                        route_remove(route, link,
-                                                     link_route_remove_handler);
+                                        route_remove(route, link, NULL);
```
Comment (ssahani) [inline src/network/networkd-dhcp4.c:217]: Why we are dropping ```link_route_remove_handler``` ?

Comment (poettering) [reply src/network/networkd-dhcp4.c:217]: we don't. It's still there, but it's not the default if the third argument is given as NULL. Hence: no code is removed, no behaviour changed, but the logic becomes a bit simpler, and nicely isolated from the outside as link_route_remove_handler can be made "static".

File: src/libsystemd/sd-netlink/netlink-util.h [hunk 2]
```diff
@@ -38,3 +38,23 @@ int rtnl_set_link_properties(sd_netlink **rtnl, int ifindex, const char *alias,
...
 int rtnl_log_create_error(int r);
+
+#define netlink_call_async(nl, ret_slot, message, callback, destroy_callback, userdata, usec, description) \
+        ({                                                              \
+                int (*_callback_)(sd_netlink *, sd_netlink_message *, typeof(userdata)) = callback; \
+                void (*_destroy_)(typeof(userdata)) = destroy_callback; \
+                sd_netlink_call_async(nl, ret_slot, message,            \
+                                      (sd_netlink_message_handler_t) _callback_, \
+                                      (sd_netlink_destroy_t) _destroy_, \
+                                      userdata, usec, description);     \
```
Comment (poettering) [inline src/libsystemd/sd-netlink/netlink-util.h]: why not go one step further here, and make the last argument `__func__` here? that would simplify the calls even further

**Comment by poettering** [comment]
I like it. Would merge it as is. Or do you want to do the __func__ thing first?

**Comment by yuwata** [comment]
@poettering Updated. `0` for timeout and `__func__` are set in the macro now. PTAL.

**Pull request merged**